File: tinyqlearning.cpp
#include<bits/stdc++.h>
#define ll long long
#define ld long double
using namespace std;
fstream f1;
ld epsilon=0.3,gamma=0.9,alpha=0.1;
vector<vector<ld>> vec(2, vector<ld>(3, 0));
vector<ll> reward = {1, 5, 2};
vector<vector<ld>> poss0 = {{0.1, 0.4, 0.5},{0.4, 0.2, 0.4},{0.3,0.7,0}};
vector<vector<ld>> poss1 = {{0.2, 0.6, 0.2},{0.5,0.25,0.25},{0.3,0.1,0.6}};
void fun1()
{
    f1.open("offlinedata.txt", ios::in);
    if(f1.is_open())
    {
        for (ll i = 0; i < 2;++i)
        {
            for (ll j = 0; j < 3;++j) f1 >> vec[i][j];
        }
        f1.close();
    }
}
void fun2()
{
    f1.open("offlinedata.txt", ios::out);
    for (ll i = 0; i < 2;++i)
    {
        for (ll j = 0; j < 3;++j) f1 << vec[i][j] <<" ";
        f1 << '\n';
    }
    f1.close();
}
int fun3(int state0,int action0)
{
    int state1;
    ld t = (rand()%100)/100.0;
    if(action0==0)
    {
        if(t<poss0[state0][0]) state1 = 0;
        else if(t<poss0[state0][0]+poss0[state0][1]) state1 = 1;
        else state1 = 2;
    }
    else
    {
        if(t<poss1[state0][0]) state1 = 0;
        else if(t<poss1[state0][0]+poss1[state0][1]) state1 = 1;
        else state1 = 2;
    }
    return state1;
}
int main()
{
    srand(time(NULL));
    fun1();
    int state0 = rand() % 3,state1=0;
    int action=0;
    cout << "begin state: " << state0 + 1 << '\n';
    cout << "        state1\tstate2\tstate3\n";
    cout << "action1 ";
    for (int j = 0; j < 2;++j) cout << vec[0][j] << '\t';
    cout <<vec[0][2]<< "\naction2 ";
    for (int j = 0; j < 2;++j) cout << vec[1][j] << '\t';
    cout <<vec[1][2]<< '\n';
    ld t=(rand()%100)/100.0;
    if(t>epsilon) action = vec[0][state0] > vec[1][state0] ? 0 : 1;
    else action = rand() % 2;
    state1 = fun3(state0,action);
    vec[action][state0] += alpha*(reward[state1] + gamma*max(vec[0][state1],vec[0][state1])-vec[action][state0]);
    state0 = state1;
    for (int i = 1; i < 100; ++i)
    {
        t=(rand()%100)/100.0;
        if(t>epsilon) action = vec[0][state1] > vec[1][state1] ? 0 : 1;
        else action = rand() % 2;
        state1 = fun3(state0,action);
        vec[action][state0] += alpha*(reward[state1] + gamma*max(vec[0][state1],vec[0][state1])-vec[action][state0]);
        state0 = state1;
    }
    fun2();
    cout<<"end state: "<<state1 + 1<<'\n';
    cout << "        state1\tstate2\tstate3\n";
    cout << "action1 ";
    for (int j = 0; j < 2;++j) cout << vec[0][j] << '\t';
    cout <<vec[0][2]<< "\naction2 ";
    for (int j = 0; j < 2;++j) cout << vec[1][j] << '\t';
    cout <<vec[1][2]<< '\n';
    return 0;
}

File: tinyqlearning.py
import random
import time
from torch.utils.tensorboard import SummaryWriter

# 初始化TensorBoard写入器
writer = SummaryWriter()
epsilon=0.3
gamma=0.9
alpha=0.1
vec=[[0.0 for _ in range(3)] for _ in range(2)]
reward=[1,5,2]
poss0=[[0.1,0.4,0.5],[0.4,0.2,0.4],[0.3,0.7,0.0]]
poss1=[[0.2,0.6,0.2],[0.5,0.25,0.25],[0.3,0.1,0.6]]
random.seed(time.time())
state0 = random.randint(0, 2)
print("begin state:",state0+1)
global_step = 0  # 全局步数计数器
def findstate(state,action0):
    t=random.random()
    if action0==0:
        if t<poss0[state][0]:
            state1=0
        elif t<poss0[state][0]+poss0[state][1]:
            state1=1
        else:
            state1=2
    else:
        if t<poss1[state][0]:
            state1=0
        elif t<poss1[state][0]+poss1[state][1]:
            state1=1
        else:
            state1=2
    return state1
def iteration(state):
    global global_step  # 使用全局步数计数器
    t=random.random()
    if(t>epsilon):
        action=0 if vec[0][state]>vec[1][state] else 1
    else:
        action=random.randint(0,1)
    state1=findstate(state,action)
    vec[action][state]+= alpha*(reward[state1] + gamma*max(vec[0][state1],vec[0][state1])-vec[action][state])
    # 记录当前步骤的所有Q值
    for i in range(2):
        for j in range(3):
            writer.add_scalar(f'Q_Value/Action_{i+1}_State_{j+1}', vec[i][j], global_step)
    
    global_step += 1  # 步数加1
    return state1
def run(state,T):
    for _ in range(0,T):
        state=iteration(state)
    print("episode=",T,",final state:",state0+1)
    print("        state1                  state2                  state3")
    print("action1", end=' ')
    for j in range(2):
        print(vec[0][j], end='\t')
    print(vec[0][2])
    print("action2", end=' ')
    for j in range(2):
        print(vec[1][j], end='\t')
    print(vec[1][2])
    return state
state0=run(state0,100)
state0=run(state0,1000)
state0=run(state0,10000)
state0=run(state0,20000)

# 关闭写入器
writer.close()